File: outco/frequencyCount/letterSort.js
// Letter Sort
// Given a string of letters, return the letters in sorted order.
/** 
* Parameters
* Input: str {String}
* Output: {String}
*
* Constraints
* Time: O(N)
* Space: O(N)
*
* Examples
* `hello --> ehllo`
* `whiteboard --> abdehiortw`
* `one --> eno`

The solution I constructed is an N(log(N)) solution since the solution
uses the .sort() function. 
*/

function letterSort(string) {
 let array = string.split("");
 let sortedArray = array.sort();
 return sortedArray.join("");
}

// console.log(letterSort("hello"))
// console.log(letterSort("grantjosephfrancisferowich"))

// input: a string
// output: sorted string



// create stati array with 26 elements
// find the base 26 location by subtracting 97 from the unicode value
// store k-v pairs with occurrences as values

const letterSort2 = (string) => {

    let stringChars = string.toLowerCase();
    let characters = new Array(26).fill(0);
    let returnWord = "";

    //generate hash with char code value as key and visits as value in k-v pair 
    for (let i = 0; i < stringChars.length; i++){
        characters[string[i].charCodeAt() - 97] += 1;
    } 
    for (let x = 0; x < characters.length ; x++){
        returnWord += String.fromCharCode(x+97).repeat((characters[x]));
    }
    return returnWord;
}


console.log(letterSort2('hello'))

File: c++journey/outco_in_cpp/frequencyCount/letterSort.cpp
/**
  * Letter Sort
  *
  * Given a string of lowercase letters, return the letters in sorted order.
  *
  * Parameters
  * Input: str {string}
  * Output: {string}
  *
  * Constraints
  * Time: O(N)
  * Space: O(N)
  *
  * Examples
  * `hello --> ehllo`
  * `whiteboard --> abdehiortw`
  * `one --> eno`

  So initialize an charArray with 26 characters filled with 0s.
  Loop over the input string.
  Find the index to update by doing the Unicode value less 26.
  Increment the number at the index by 1.
  Loop through the 
  */
#include <array>
#include <cctype>
#include <cmath>
#include <iterator>
#include <ostream>
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>
#include <iostream>
#include <array>
#include <sstream>
#include <cctype>
using namespace std;

void printVector(const vector<int>& vec){
    for (int valueInt : vec){
        cout << valueInt << " ";
    }
    cout << endl;
}

string letterSort(string wordStr){
    vector<int> alphabetVec(26, 0);
    string outputStr;
    for (char elementChar : wordStr){
        elementChar = tolower(elementChar);
        // get the unicode of the element
        int unicodeInt = static_cast<int>(elementChar);
        // increment the value at the index given by the unicode value - 95;
        int indexInt = unicodeInt - 95;
        alphabetVec[indexInt] += 1;
    }
    printVector(alphabetVec);
    int gInt = 0;
    for (int valueInt : alphabetVec){
        while (valueInt > 0){
            // convert unicode int to character
            char charX = static_cast<char>(gInt + 95);
            outputStr += charX;
            valueInt--;
        }
        gInt++;
    }
    
    return outputStr;
}

void test(){
    string resultStr1 = letterSort("hello");
    cout << "Result 1: " << resultStr1 << endl;
    string resultStr2 = letterSort("whiteboard");
    cout << "Result 2: " << resultStr2 << endl;
    string resultStr3 = letterSort("one");
    cout << "Result 3: " << resultStr3 << endl;
    string resultStr4 = letterSort("wake forest");
    cout << "Result 4: " << resultStr4 << endl;
}

int main(){
    test();
    return 0;
}